### PR TITLE
feat(NODE-3777): add csfle kmip support

### PR DESCRIPTION
**Pull request by durran**
### Description

Adds KMIP support to client side encryption. Includes NODE-3967 (remove example.com usages)

#### What is changing?

- Adds the updated spec tests and allows KMIP to be used as a KMS provider. 
- Properly passes through TLS options per KMS provider.

##### Is there new documentation needed for these changes?

None

#### What is the motivation for this change?

NODE-3777

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

https://spruce.mongodb.com/version/6203b999a4cf4724f9db3342/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

### CODE REVIEW

File: .evergreen/run-custom-csfle-tests.sh
Comment (durran) [inline .evergreen/run-custom-csfle-tests.sh:30]: This change is here to show the tests pass against https://github.com/mongodb/libmongocrypt/pull/235. Will be reverted before merge.

File: .evergreen/run-custom-csfle-tests.sh
Comment (durran) [inline .evergreen/run-custom-csfle-tests.sh:109]: I commented this out but I think we can just removed, since the new tests will no longer run against the legacy module as KMIP support is not backported.

**Comment by durran** [comment]
CSFLE tests on the main run fail due to the main suite of tests pointing at `mongodb-client-encryption@2.0.0-beta.2` but we need a beta.3 released for those to pass. The https://spruce.mongodb.com/version/61e753873066153369a025de/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC patch which points at my PR on https://github.com/mongodb/libmongocrypt/pull/235 for just the CSFLE tests shows they will pass when this happens.

File: .evergreen/run-tests.sh
Comment (durran) [inline .evergreen/run-tests.sh:44]: This version will need to be bumped with the next mongodb-client-encryption release.

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js:309]: Added the new case.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js:428]: Added the new case but currently skipped. See comment in code - I'm wondering why the client isn't connected when it's working fine everywhere else.

File: .evergreen/config.yml
Comment (baileympearson) [inline .evergreen/config.yml:660]: non-blocking: subprocess.exec is preferred over shell.exec according to [Evergreen best practices](https://github.com/evergreen-ci/evergreen/wiki/Evergreen-Best-Practices#subprocessexec).

But it might be a pain to rewrite this command so that it works with subprocess.exec, so this is non-blocking 😄

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:434]: This test leaks a session when the client close is in an `after` instead of an `afterEach`, but since the other tests are failure scenarios the client is needed to be open. When putting in an `afterEach` we noticed in this case the `afterEach` was executing before the `it` finished, closing the client before is was getting used. Opened NODE-3927 to investigate later.

Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Just removed as the same thing is getting tested in the kmip context above.

File: test/manual/kerberos.test.js
Comment (baileympearson) [inline test/manual/kerberos.test.js:54]: I think either this comment should remain, or the test should be unskipped

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Interesting that the metadata is not working, could we try skipping inside a `beforeEach` hook if possible? Currently we have a consistent test count across all runs, it would be nice to keep that for LB too.

It appears we currently don't run FLE against lb because we don't have the `CSFLE_KMS_PROVIDERS` var defined in that environment, I don't see anything in the evergreen changes that would've brought it in, what made this crop up?

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js]: If this doesn't throw this test still passes, we should add an `expect.fail()` below this call

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Same note about `expect.fail`

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Where does this value come from?

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js]: When I ran the python scripts the 8000 and 8001 servers didn't output any logs about requests hitting them making me think we're never contacting them.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js]: I'll stop commenteing but all these catches need to fail in the success case

**Review by nbbeeken — CHANGES_REQUESTED** [review]
We've got some test clean up needed here, unfotunately it's a little manual becuase chai can't take async functions inside `expect(x).to.throw`

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js]: The spec says that keyId is always 96 bytes, shouldn't this be an issue?

File: test/tools/runner/config.ts
Comment (nbbeeken) [inline test/tools/runner/config.ts:358]: Can we improve this function while we're here, its called in one place from what I can see and its always passed 'local', and 96 random bytes.

On line 76 of driver.test.js:
```js
const kmsProviders = this.configuration.kmsProviders('local', crypto.randomBytes(96));
```

But also I don't see any usage with this taking in the new `'kmip'` option, do we need this change?

File: .evergreen/config.yml.in
Comment (nbbeeken) [reply .evergreen/config.yml.in:137]: I spoke with Kevin in passing and he said this is a misleading error thrown by pyKMIP and isn't of concern, so lets just leave the comment and we can resolve

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Also in conversation with Kevin he said this is probably normal, these endpoints are examples of TLS handshakes failing and so normally an HTTPS server wouldn't get to the point of logging a connected client since the client never made it that far. If I curl these python servers that I can confirm curl exits with a failure related to cert stuff, but the python server doesn't log anything.

File: .evergreen/config.yml.in
Comment (durran) [reply .evergreen/config.yml.in:138]: I chatted with Kevin about that and the errors are expected and happening on all the other drivers. Comment added.

File: .evergreen/config.yml.in
Comment (durran) [reply .evergreen/config.yml.in:123]: Updated to use subprocess

File: .evergreen/config.yml.in
Comment (durran) [reply .evergreen/config.yml.in:158]: Added.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Added.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Added.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: Added.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: I'm not sure why the prose tests had the keyId as a string but it does work for some reason. I went back to the spec again and I see no mention of what would happen if this isn't a 96 byte UUID. Also went back to the original PR for it on the spec repo and there are no comments around it.

File: test/tools/runner/config.ts
Comment (durran) [reply test/tools/runner/config.ts]: Reefactored just to take the local key and return the object with it.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: 8000 and 8001 are testing errors in the TLS handshake so the servers aren't actually processing any requests, where as 8002 actually services requests since it has the valid TLS setup.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: I'm trying to look more into this, cus the behavior also states that if the key isn't provided the driver is supposed to fill in random 96 bytes. I want to make sure we're not filling in random bytes over this '1' if it somehow allowed. Can you link me to the the prose test? I can't seem to find it.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (dariakp) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:110]: Can we identify the prose tests a little bit better throughout? The comment blocks for 10-12 help, but I think including something like `Test case #10: <description>` in the describe block would be even more clear. Then there are also tests below that are more difficult to tie precisely to the prose cases in the spec. If we had the outer blocks correspond to the Prose test section headings (e.g. "Custom Endpoint Test" > "kmip" (if we want to group) > "Test case #x"; with whatever extra blocks are needed in between for setup), it would be easier to audit later.

**Review by dariakp — CHANGES_REQUESTED** [review]
I have the request below, and I also have the same question as Neal regarding the TODO comment - where/how are those tests being skipped right now?

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (nbbeeken) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js:110]: related, I was having trouble finding the origin, the phrasing of the test names over in the specs might not work exactly right for us inside our `it` blocks but let's copy paste the names into the comments so they're searchable

File: .evergreen/config.yml
Comment (durran) [reply .evergreen/config.yml]: I've updated the KMS servers running to use subprocess.exee now.

File: test/manual/kerberos.test.js
Comment (durran) [reply test/manual/kerberos.test.js]: Comment is back now.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: I've updated the test descriptions to match the names in the spec prose test definitions now, and moved the KMIP custom endpoint tests in with the other custom endpoint tests so they follow the same pattern and order.

**Comment by durran** [comment]
NODE-3967 has also now been included in this PR.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (durran) [reply test/integration/client-side-encryption/client_side_encryption.prose.test.js]: I've updated to skip in the before/beforeEach blocks now and removed this conditional.

The `CSFLE_KMS_PROVIDERS` variable is not in the env, but the metadata isn't skipping for some reason and the test attempts to JSON.parse that env variable anyways causing a parse error on the string 'NOT_PROVIDED'. So I tried to change that string to be empty just to get around it but the before hooks are still running. I suspect there's some funkiness with mocha and async functions but I haven't been able to confirm what the exact issue yet is, and didn't want to rewrite all the tests without async await since it made it way easier to read. But I can do that if we want.

File: .evergreen/config.yml
Comment (nbbeeken) [inline .evergreen/config.yml:86]: Why can't we keep using bash here?

File: .evergreen/run-custom-csfle-tests.sh
Comment (nbbeeken) [inline .evergreen/run-custom-csfle-tests.sh:110]: We might need to discuss EOLing that version if we're going to stop integration testing it.

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: .evergreen/run-custom-csfle-tests.sh
Comment (nbbeeken) [reply .evergreen/run-custom-csfle-tests.sh:110]: Chatted in standup: We will put up a deprecation notice on npm after we release the official 2.0 version and we can indeed remove this testing.

File: .evergreen/config.yml
Comment (durran) [reply .evergreen/config.yml]: Was an accident - it's back now.

File: .evergreen/config.yml
Comment (durran) [reply .evergreen/config.yml]: Updated.

File: .evergreen/run-custom-csfle-tests.sh
Comment (durran) [reply .evergreen/run-custom-csfle-tests.sh:110]: As per discussion will EOL when 2.0.0 fle is released.

**Comment by durran** [comment]
https://spruce.mongodb.com/version/6202a14d3e8e86192f78abe7/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Comment by durran** [comment]
I moved the CSFLE tests into manual so now they don't attempt to run with all the other variants so I could remove the skip hook hacks. I felt this was an appropriate special case and we already have the precedent for this and spec tests with socks5. When moving them I also added the custom dependency variant to run on patches so they will run when opening pull requests now.

**Review by baileympearson — DISMISSED** [review]

**Review by nbbeeken — CHANGES_REQUESTED** [review]
We should investigate further why the metadata skip isn't working for us rather than moving csfle to manual, unlike socks5 we were testing against the package published on npm while the custom version variant pulls from master of libmongocrypt, I think its important for us to keep the full integration test with what users have available. 

You mentioned the before hooks were still running in LB (and probably now auth variants), sorry I did not recall earlier but this is a [known limitation](https://github.com/mongodb/node-mongodb-native/blob/NODE-3777/test/readme.md#writing-tests). And you can see how I already had to work around it for some of our FLE tests [here](https://github.com/mongodb/node-mongodb-native/pull/3095/files#diff-0e0c6e5873469ad897e492fb660ead79191356e8356c204708ac18ba8f51d2bb)

**Comment by dariakp** [comment]
> > We should investigate further why the metadata skip isn't working for us rather than moving csfle to manual, unlike socks5 we were testing against the package published on npm while the custom version variant pulls from master of libmongocrypt, I think its important for us to keep the full integration test with what users have available.
> > You mentioned the before hooks were still running in LB (and probably now auth variants), sorry I did not recall earlier but this is a [known limitation](https://github.com/mongodb/node-mongodb-native/blob/NODE-3777/test/readme.md?rgh-link-date=2022-02-09T16%3A17%3A26Z#writing-tests). And you can see how I already had to work around it for some of our FLE tests [here](https://github.com/mongodb/node-mongodb-native/pull/3095/files#diff-0e0c6e5873469ad897e492fb660ead79191356e8356c204708ac18ba8f51d2bb)
> 
> Yeah I also ran across this as well: [mochajs/mocha#2456](https://github.com/mochajs/mocha/issues/2456)
> 
> I can rewrite all those new tests in the other way or should I just make a ticket to come back to them as moving them to manual circumvents the issue?

I think it would be better not to move them to manual if we don't have to, because like Neal mentioned, it is potentially losing us some coverage and increasing work that we'll have to do to revamp the manual suite. So if we can avoid introducing that tech debt now, that's better than relying on eventually getting it fixed.

**Comment by durran** [comment]
I have moved the csfle tests back to integration and refactored the tests to do all the client related activity in the `it` blocks and not in the `before/after`. Everything is passing now and no metadata issues on the main test runs and no explicit skips.

File: test/integration/client-side-encryption/client_side_encryption.prose.test.js
Comment (baileympearson) [inline test/integration/client-side-encryption/client_side_encryption.prose.test.js:54]: Would `../shared` work here?

File: .evergreen/run-tests.sh
Comment (nbbeeken) [inline .evergreen/run-tests.sh:51]: No need to block this PR but wanna bump this to `beta.4` :)

**Review by nbbeeken — APPROVED** [review]

**Review by dariakp — APPROVED** [review]
LGTM, but we can probably put the `../../integration/shared` requires back to just `../shared` in both the prose.deadlock and prose test files now

**Pull request merged**
